DataManager.cpp

```diff
@@ -702,28 +700,11 @@ void DataManager::resumeRemoteChunk() {
   delete currentChunkBuffers->cb;
   delete currentChunkBuffers;
 
-#ifdef PINNED_HOST_MEMORY
-  if(bufRemoteMoments != NULL)
-      freePinnedHostMemory(bufRemoteMoments);
-  if(bufRemoteParts != NULL)
-      freePinnedHostMemory(bufRemoteParts);
-#else
-  if(bufRemoteMoments != NULL)
-      free(bufRemoteMoments);
-  if(bufRemoteParts != NULL)
-      free(bufRemoteParts);
-#endif
-
-    // resume each treepiece's startRemoteChunk, now that the nodes
-    // are properly labeled and the particles accounted for
-    for(int i = 0; i < registeredTreePieces.length(); i++){
-      if(verbosity > 1) CkPrintf("[%d] resumeRemoteChunk %d\n", CkMyPe(), i);
-      int in = registeredTreePieces[i].treePiece->getIndex();
-#if COSMO_PRINT_BK > 1
-      CkPrintf("(%d) dm->%d\n", CkMyPe(), in);
-#endif
-      treePieces[in].continueStartRemoteChunk(chunk, (intptr_t)d_remoteMoments, (intptr_t)d_remoteParts);
-    }
+ // Check and see if the remote walks already finished and are waiting
+ // to launch their GPU kernels
+ for (int i = 0; i < numPEListProxies; i++) {
+     PEListProxies[i]->tryLaunchDelayedKernel();
```

Comment:
This doesn't seem to be doing what I want. There are 6 PEList proxies, so I need 6 invocations of this function on each PE.

When I run on a single node with 2 ranks, I'm seeing this function get called 12 times on each PE. Any idea what's going on here? Is there a better way to do this?

Reply:
I was wondering if this was possible. I imagine you mean something like

```
 int pe;
 int firstPE = CkNodeFirst(CkMyNode());
 int nPEs = CkNodeSize(CkMyNode());
 for (int i = 0; i < numPEListProxies; i++) {
     for (int j = 0; j < nPEs; j++) {
         pe = firstPE + j;
         PEListProxies[i][pe].tryLaunchDelayedKernel();
     }
 }
```

The function call on the proxy is giving me a seg fault here. It's not clear to me if ordering is guaranteed here.

Reply:
The other way I was thinking of doing this was to maintain a 'registeredPEList' array in the DataManager. Similar to the 'registeredTreePieces', I'm wondering if the 'thisIndex' property is necessary to iterate over the elements of the PE group like this.

Reply:
Oh right, I forgot that I made PEListProxies an array of pointers. The intention there was to make the sendList() calls in Compute.cpp easier to read.

Your suggestion appears to work. Although I still haven't seen any cases where the remote walk beats the remote data transfer and triggers the protection against this race condition.